packages/good-design/src/core/modals/BasicModal.tsx

```diff
@@ -0,0 +1,76 @@
+import { Button, Modal } from "native-base";
```

Comment:
i'm a bit sceptical about such schema (hook returning component).
hooks & components are things existing in parallel. hooks are about logic, components about presentation
(in the ideal case of course). but here the're twisted too much, form my point of view

here's an alternate solution i could suggest. it could be made a bit simpler by using some module-level global var instead of the context:


```typescript
import { noop } from "lodash"
import { v4 as uuidv4 } from "uuid"
import { Button, Modal as NativeModal } from "native-base";
import React, { createContext, useCallback, useState, useRef, useEffect, useContext } from "react";

// Context could be replaced with
// const state = {}
// const toggleState = (id: string, value: boolean) => state[id] = value

const ModalContext = createContext({
  state: {},
  toggleState: (id: string, state: boolean) => {},
});

const ModalContextProvier = ({ children }) => {
  const [state, setState] = useState({})

  const toggleState = useCallback((id: string, state: boolean) => {
    setState(oldState => ({ ...oldState, [id]: state }))
  }, [setState])

  return (
    <ModalContext.Provider value={{ state, toggleState }}>
      {children}
    </ModalContext.Provider>
  )
}

// called without args will generate new unique ID to be passed down to the modal component
// called with specific id will return functions managing state of modal with this exact id
const useModal = (id = null) => {
  const [identifier, setIdentifier] = useState(() => id || uuidv4());
  const { state, toggleState } = useContext(ModalContext)

  const showModal = useCallback(() => toggleState(identifier, true), [toggleState, identifier])
  const hideModal = useCallback(() => toggleState(identifier, false), [toggleState, identifier])

  useEffect(() => {
    setIdentifier(id || uuidv4())
  }, [setIdentifier, id])

  return { modalId: identifier, modalShown: state[identifier], showModal, hideModal }
}

const Modal = ({
  modalId,
  header,
  body,
  footer,
  actionText,
  closeText = "Cancel",
  onClose = noop,
  onAction = noop,
  _modal,
  _header,
  _body,
  _footer
}) => {
  const { modalShown, hideModal } = useModal(modalId)

  const onModalAction = useCallback(() => {
    hideModal()
    onAction()
  }, [hideModal, onAction])

  const onModalClose = useCallback(() => {
    hideModal()
    onClose()
  }, [hideModal, onClose])

  return (
    <NativeModal isOpen={modalShown} onClose={onModalClose} {..._modal}>
      <NativeModal.Content>
        {closeText && <NativeModal.CloseButton />}
        <NativeModal.Header {..._header}>{header}</NativeModal.Header>
        <NativeModal.Body {..._body}>{body}</NativeModal.Body>
        <NativeModal.Footer {..._footer}>
          {footer}
          {(actionText || closeText) && (
            <Button.Group space={2}>
            {closeText && (
              <Button
                variant="ghost"
                colorScheme="blueGray"
                onPress={onModalClose}
              >
                {closeText}
              </Button>
            )}
            {actionText && (
              <Button onPress={onModalAction}>
                {actionText}
              </Button>
            )}
            </Button.Group>
          )}
        </NativeModal.Footer>
      </NativeModal.Content>
    </NativeModal>
  );
}

// usage example
// useSignWalletModal.tsx -> SignWalletModal.tsx
const SignWalletModal = ({ txStatus, ...props }) => {
  const { modalId, showModal, hideModal } = useModal()

  useEffect(() => {
    if (["PendingSignature", "CollectingSignaturePool"].includes(txStatus as any)) {
      showModal();
    } else if (txStatus) {
      hideModal();
    }
  }, [txStatus]);

  return (
    <Modal
      modalId={modalId}
      body={"Please sign the transaction in your wallet..."}
      closeText=""
      {...props}
    />
  );
}


```

Reply:
what do you mean skeptical?
1. either it works or not
2. dont fix it or refactor it if it is not broken
3. definitely DO NOT waste time on writing a complete suggestion on how to refactor. This is the same as spending time refactoring it! and means you are ignore rule 2!

Reply:
@sirpy we need to discuss this on tech sync
no one mixes hooks & components this way.
I haven't seen this in any library source 
so we already **have broken design**

also, in my solution it's more easy to extend/make custom popups

Reply:
@sirpy react provides a lot of ready to use solution and proper design patterns 
personally, before start some feature I;m checking for any generic ways to do it
I research what React could provide/suggest. and only if nothing - I'm implementing the custom solution